docs/src/references/mappers/gen-mapper.md

```diff
@@ -0,0 +1,121 @@
+---
+title: Generic Mapper
+tags: [Reference, Mappers, Cloud]
+sidebar_position: 2
+draft: true
+---
+
+import ProposalBanner from '@site/src/components/ProposalBanner'
+
+<ProposalBanner/>
+
+:::note
+This section is actually a design document.
+It includes a reference guide for the POC, but also proposes a plan toward a generic mapper.
+:::
+
+## Motivation
+
+In theory, %%te%% users can implement customized mappers to transform data published on the MQTT bus
+or to interact with the cloud. In practice, they don't.
+Implementing a mapper is costly while what is provided out-the-box by %%te%% already meets most requirements.
+The need is not to write new mappers but to adapt existing ones.
+
+The aim of the generic mapper it to let users extend and adapt the mappers with their own filtering and mapping rules,
+leveraging the core mapping rules and mapper mechanisms (bridge connections, HTTP proxies, operations).
+
+## Vision
+
+The %%te%% mappers for Cumulocity, Azure, AWS and Collectd are implemented on top of a so-called generic mapper
+which is used to drive all MQTT message transformations.
+- Transformations are implemented as pipelines consuming MQTT messages, feeding a chain of filters and producing MQTT messages.
+  - `MQTT sub| filter-1 | filter-2 | ... | filter-n | MQTT pub`
+- A pipeline can combine builtin and user-provided filters.
+- The user can configure all the transformations used by a mapper,
+  editing MQTT sources, pipelines, filters and MQTT sinks.
+- By contrast with the current implementation, where the translation of measurements from %%te%% JSON to Cumulocity JSON
+  is fully hard-coded, with the generic mapper a user can re-use the core of this transformation while adding customized steps:
+  - consuming measurement from a non-standard topic
+  - filtering out part of the measurements
+  - normalizing units
+  - adding units read from some config
+  - producing transformed measurements on a non-standard topic.
+
+## POC reference
+
+- The generic mapper loads pipeline and filters stored in `/etc/tedge/gen-mapper/`.
+- A pipeline is defined by a TOML file with `.toml` extension.
+- A filter is defined by a Javascript file with `.js` extension.
+- The definition of pipeline must provide a list of MQTT topics to subscribe to.
+  - The pipeline will be feed with all the messages received on these topics.
+- A pipeline definition also provides a list of stages.
+  - Each stage is built from a javascript and is possibly given a config (arbitrary json that will be passed to the script)
+  - Each stage can also subscribe to a list of MQTT topics (which messages will be passed to the script to update its config)
+
+```toml
+input_topics = ["te/+/+/+/+/m/+"]
+
+stages = [
+    { filter = "add_timestamp.js" },
+    { filter = "drop_stragglers.js", config = { max_delay = 60 } },
+    { filter = "te_to_c8y.js", meta_topics = ["te/+/+/+/+/m/+/meta"] }
+]
+```
+
+- A filter has to export at least one `process` function.
```

Comment:
Having the function signature also defined here or prior to this section would be good.